third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs

```diff
@@ -0,0 +1,381 @@
+use move_stackless_bytecode::{
+    dataflow_analysis::{DataflowAnalysis, TransferFunctions},
+    dataflow_domains::{AbstractDomain, JoinResult},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::Bytecode,
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::collections::BTreeMap;
+
+/// State of initialization of a local at a given program point.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Initialized {
+    No,    // definitely not initialized
+    Maybe, // maybe initialized
+    Yes,   // definitely initialized
+}
+
+impl AbstractDomain for Initialized {
+    /// Implements `join` for the initialized state lattice:
+    ///
+    ///             +-------+
+    ///             | Maybe |
+    ///             +-------+
+    ///               /   \
+    ///              /     \
+    ///      +-----+/       \+----+
+    ///      | Yes |         | No |
+    ///      +-----+\       /+----+
+    ///              \     /
+    ///               \   /
+    ///            +--------+
+    ///            | bottom |
+    ///            +--------+
+    /// Note that bottom is not explicitly represented in the code.
+    fn join(&mut self, other: &Self) -> JoinResult {
+        if *self == *other {
+            return JoinResult::Unchanged;
+        }
+        if *self != Initialized::Maybe {
+            *self = Initialized::Maybe;
+            JoinResult::Changed
+        } else {
+            JoinResult::Unchanged
+        }
+    }
+}
+
+/// Initialization state of all the locals at a program point.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct InitializedState(Vector<Initialized>);
+
+impl InitializedState {
+    /// Create a new initialized state with:
+    /// * all param locals set to `Yes`
+    /// * all other locals set to `No`.
+    /// Note: `num_locals` is the total number of locals, including params.
+    pub fn new(num_params: usize, num_locals: usize) -> Self {
+        if num_locals < num_params {
+            panic!("ICE: num_locals must be >= num_params");
+        }
+        Self(Vector::from_iter(
+            std::iter::repeat(Initialized::Yes)
+                .take(num_params)
+                .chain(std::iter::repeat(Initialized::No).take(num_locals - num_params)),
+        ))
+    }
+
+    /// Mark `local` as initialized.
+    fn mark_as_initialized(&mut self, local: usize) {
+        self.0.set(local, Initialized::Yes);
+    }
+
+    /// Get the initialization state of `local`.
+    /// Panics if `local` does not exist in this state.
+    fn get_initialized_state(&self, local: usize) -> Initialized {
+        self.0.get(local).expect("local must exist").clone()
+    }
+}
+
+impl AbstractDomain for InitializedState {
+    /// `join` of two initialized states is the point-wise join for each local.
+    /// The result is `JoinResult::Changed` if any of the local joins causes a change.
```

Comment:
This seems also a very common situation, where we each local has an abstract value from a lattice T, and the abstract state of a program point is T^n where the program n locals. In code: `impl<T: AbstracDomain> AbstrctDomain for Vec<T>`

Reply:
https://github.com/aptos-labs/aptos-core/issues/11632

Reply:
Note that I am using a `im::Vector` here, but your general point still stands. I think it might be a good exercise to do some of these abstractions as a common refactoring pass. Given that this PR is already quite large, I think it makes sense to do this in a different pass: thank you for filing an issue to track this.